client/src/connection_cache.rs

```diff
@@ -276,48 +317,79 @@ fn get_or_add_connection(addr: &SocketAddr) -> GetConnectionResult {
             (connection.clone(), true, map.stats.clone(), stats, 0, 0)
         }
         None => {
-            let (_, send_socket) = solana_net_utils::bind_in_range(
-                IpAddr::V4(Ipv4Addr::new(0, 0, 0, 0)),
-                VALIDATOR_PORT_RANGE,
-            )
-            .unwrap();
-
-            let connection = if map.use_quic {
-                Connection::Quic(Arc::new(QuicTpuConnection::new(send_socket, *addr)))
-            } else {
-                Connection::Udp(Arc::new(UdpTpuConnection::new(send_socket, *addr)))
-            };
-
-            // Upgrade to write access by dropping read lock and acquire write lock
-            drop(map);
+            // Upgrade to write access by dropping read lock and acquire write lock on the connection
+            // the lock on the connection map is still readonly -- as we only need to read it until we
+            // add a connection entry to it
+
             let mut get_connection_map_lock_measure =
                 Measure::start("get_connection_map_lock_measure");
-            let mut map = (*CONNECTION_MAP).write().unwrap();
+            drop(addr_read_lock);
```

Comment:
Why are we dropping this lock here, when I don't see anywhere where we do anything with it? We still acquire an RwLock on the ConnectionMap, so I'm kind of confused about the purpose of CONNECTION_LOCK_MANAGER.

Reply:
We drop the read lock on the connection for address -- it will be escalated to write lock. The write lock ensure the connection per address is only created once per time.

Reply:
Connection lock manager maintains the map of locks -- from SocketAddr --> Mutex. This way different addresses are not contending with each other

Reply:
See line 328

Reply:
Why do we drop(map) and relook it is to maintain the locking order.

The order is LockManager Lock Per Address First followed by CONNECTION_MAP lock. This is required to ensure no deadlocks.

Reply:
_addr_write_lock still seems unused though, and we're still acquiring a write lock on CONNECTION_MAP. Apologies for the confusion, but can you please clarify why we need to read-lock or write-lock the RwLocks coming from CONNECTION_LOCK_MANAGER or how this gets rid of the lock contention on CONNECTION_MAP?

Reply:
It is used -- it preventing connection from being made for the attempt with the same address. It is doing its work until it drops out of scope

Reply:
We acquire write lock on CONNECTION_MAP only at the last minute when we really need make data structure change to it. Notice we are not acquiring it for write when calling QuicTpuConnection::new

Reply:
To more easily understand: CONNECTION_LOCK_MANAGER ensure one actual connection is made per address a time, CONNECTION_MAP lock ensure its data structure integrity.

Reply:
The CONNECTION_MAP Lock handling could be better encapsulated in ConnectionMap itself. But It would be for another PR.

Reply:
The story looks like the following
1. When get_or_add_connection is called for an address, I lock the address for read, and lock the connection for read as well and check if it is in the cache. If in the cache, happily return the existing connection. Otherwise,
2. We did not find the connection in the cache, so potentially we are going to create a new connection for it. We will drop the two locks in 1, then I acquire the address lock for write to ensure no one else for the same address is going to create a connection. The connection map lock also acquired for read -- because we are checking the cache: read. After we do this, we checking the cache again, as it is possible some one else raced with us that when we drop our locks and require them, a connection with the same address is already created. That is why you see the double read of the connection map. If we find in the cache, we drop all the locks and happily return the connection for this address. Otherwise,
3. We have acquired the write lock on the address, and did not find in the cache, it means we have to create it, and we are the only one being able create it due to the write lock for that address. But other thread working on a different address, can still come in here to create other connections as the CONNECTION_MAP lock is still acquired shared. Continue
4. After the connection is created for the address, we are going to add it to the CONNECTION_MAP, to ensure its data structure's integrity, we are going to lock the CONNECTION_MAP exclusively. After it is added, both locks are dropped.

Reply:
Ok... sort of makes a bit more sense... would it be possible to actually get rid of the CONNECTION_MAP entirely and just have the entries in CONNECTION_LOCK_MANAGER be RwLock<Option<Connection>>? I think that way it would make more sense and maybe be slightly faster.

Reply:
We cannot get rid of protection on CONNECTION_MAP all together. But we can move the locking logic inside (lock shared when we need to read , lock exclusive when we need to modify) pretty locally. So it will be easier to understand. Notice I have the internal lock for LockManager hidden in it -- you do not care about it. The idea will be the same. I plan to do it in a follow on PR to limit the change size if you are Okay.

Reply:
I think it's possible to just have the Connections protected by an individual RwLock<Option\<Connection\>> owned by CONNECTION_LOCK_MANAGER, so we could get rid of CONNECTION_MAP entirely.

Reply:
So a definition like

`#[derive(Default)]
pub struct LockManager<T>
where
    T: std::cmp::Eq + std::hash::Hash,
{
    locks: Mutex<HashMap<T, Arc<RwLock<Option<Connection>>>>>,
}`

Reply:
The LockManager is intended a little generic -- so that it could be used for other use cases hence the <T> :) -- notice the LockManager would have to mange the caching and eviction stuff. So it does not have other special logic than the just map of RwLocks. I think the alternative is copying the idea of LockManager and get rid of the LockManager itself and let ConnectionMap handle all of this.